recursion.md

```diff
@@ -0,0 +1,172 @@
+achieve in a recursive solution.
+
+```rb
+return 1 if exp == 0
+return base if exp == 1
+return base * base if exp == 2
+```
+
+If you structure your problem right, all you need for your base case is the
+first one - `return 1 if exp == 0`.
+
+2) Be clear and consistent in terms of what's the return value of the recursive algorithm.
+
+## Exponent
+
+Write a function that takes a number (base) and an exponent. Raise the base to
+the power of a given exponent.
+
+**Solution 1: Simplest recursive solution - positive numbers**
+
+This is perhaps the most straight-forward way to solve the exponent/to the power
+of problem using recursion. Note that this first solution will only work for
+positive exponents.
+
+```rb
+def power(base, exponent)
+  if exponent == 0
+    return 1
+  else
+    power(base, exponent - 1)
+  end
+end
+```
+
+Here's an example of what the callstack would look if we call `power(2,3)`
+
+exponent|recursive call               |return value
+--------|-----------------------------|------------
+0       |                             | 1
+1       |2 * power(2, (1-1)) => 2 * 1 | 2
+1       |2 * power(2, (2-1)) => 2 * 2 | 4
+2       |2 * power(2, (2-1)) => 2 * 4 | 8
+
+Runtime - O(n):
+
+Essentially, what we're trying to find the cost of exponent ``(exponent - 1)`` + the base case`( 0(1))`. This means that our runtime complexity is `O(exponent)`.
+https://cs.muic.mahidol.ac.th/courses/ds/lects/lect03.pdf
+
+Size - O(n): Only n number of stacks required at any one time.
+
+To understand recursion, let's also take look at an iterative solution.
+
+**Solution 2: Simple iterative solution**
+
+```rb
+def power(base, exponent)
+  result = 1.0
+  if exponent > 0
+    exponent.times do |_|
+      result *= base
+    end
+  else
+    (-1 * exponent).times do |_|
+      result /= base
+    end
+  end
+
+  result
+end
+```
+
+Notes on negative powers:
+
+1) Dealing with negativity: One of the easiest ways to deal with a negative
+   exponent is to make it positive and have it divided by 1.
+2) Dealing with decimals. If we want to find `power(2, -2)`, then our algorithm   
+   will run 1/4 as its final operation. In Ruby this will return 0, but the
+   answer we're looking for is `0.25`. Therefore using `1.0` ensures that we'll
+   return be able to return the correct decimal number (  a float).
+
+Runtime - O(n):  
+Size - O(n): Only n number of stacks required at any one time.
+
+**Solution 3: More efficient solution - positive numbers**
+
+By only making recursive calls for the exponent / 2, we can halve the size of
+our call stack and halve the size of our problem space at each call.
+
+```rb
+def power(base, exponent)
+  if exponent == 0
+    return 1
+  elsif exponent % 2 == 0
+    half = power(base, exponent / 2)
+    return half * half
+  else
+    half = power(base, exponent / 2)
+    return base * half * half
+  end
+end
+```
+
+Runtime - O(log(n)): For every n in 1..exponent, we perform n/2
```

Comment:
It seems like this problem is logarithmic because every other call divides the problem space in 2. However, these calls run n/2 operations each times so that feels slightly greater than O(log(n)). Thoughts?

Reply:
What do you mean "these calls run n/2 operations each time"? This seems to be a classic case of logarithmic time.

In the course of the run time of the program, there are three possibilities. Either an immediate return (base case). The two other possibilities occur exclusively (either one occurs or the other occurs but not both) and whenever one occurs, the problem space is divided in half

Reply:
Yeah, I think my confusion in the previous comment about why the runtime was `O(exponent)` was what continued here - for some reason I had it in my head that these recursive calls were somehow nested. In fact, they are actually just acting like a (linear) loop control structure (or in this case, a logarithmic one)